### PR TITLE
Delay ambiguous intra-doc link resolution after `Cache` has been populated

### LINKED ISSUE #130233: Intra-doc link should rule out hidden items when resolving ambiguous link
```rust
pub struct Thing {}

#[doc(hidden)]
#[allow(non_snake_case)]
pub fn Thing() {}

/// Do stuff with [`Thing`].
pub fn repro(_: Thing) {}
```

```console
$ cargo doc

warning: `Thing` is both a function and a struct
 --> src/lib.rs:7:21
  |
7 | /// Do stuff with [`Thing`].
  |                     ^^^^^ ambiguous link
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
help: to link to the function, add parentheses
  |
7 | /// Do stuff with [`Thing()`].
  |                          ++
help: to link to the struct, prefix with `struct@`
  |
7 | /// Do stuff with [`struct@Thing`].
  |                     +++++++
```

I expected to see this happen: No broken_intra_doc_links warning. ```[`Thing`]``` works and means the same thing as ```[`struct@Thing`]```, because ```[`Thing()`]``` has been ruled out for being hidden. Explicitly writing ```[`Thing()`]``` produces a broken link. Only when documented with `--document-hidden-items` is there an ambiguity.

Instead, this happened: ```[`Thing`]``` and ```[`Thing()`]``` each produce broken link.

### Meta

`rustc --version --verbose`:
```
rustc 1.83.0-nightly (0ee7cb5e3 2024-09-10)
binary: rustc
commit-hash: 0ee7cb5e3633502d9a90a85c3c367eccd59a0aba
commit-date: 2024-09-10
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust/issues/130233.

I was getting nowhere with #130278. I took a wrong turn at some point and ended making way too many changes so instead I started again back from 0 and this time it worked out as expected.

r? @notriddle

### CODE REVIEW

**Review by notriddle — CHANGES_REQUESTED** [review]

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -243,18 +247,63 @@ struct ResolutionInfo {
...
+    ///
+    /// So instead, we store the ambiguous links and we wait for cache paths to be filled before
+    /// inferring them (if possible).
+    ///
+    /// Key is `(item ID, path str)`.
+    pub(crate) ambiguous_links: FxIndexMap<(ItemId, String), Vec<AmbiguousLinks>>,
+}
+
+pub(crate) struct AmbiguousLinks {
+    disambiguator: Option<Disambiguator>,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:303]: One word: reexports. :)

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -243,18 +247,63 @@ struct ResolutionInfo {
...
     /// Cache the resolved links so we can avoid resolving (and emitting errors for) the same link.
     /// The link will be `None` if it could not be resolved (i.e. the error was cached).
-    visited_links: FxHashMap<ResolutionInfo, Option<(Res, Option<UrlFragment>)>>,
+    pub(crate) visited_links: FxHashMap<ResolutionInfo, Option<(Res, Option<UrlFragment>)>>,
+    /// These links are ambiguous. We need for the cache to have its paths filled. Unfortunately,
+    /// if we run the `LinkCollector` pass after `Cache::populate`, a lot of items that we need
+    /// to go through will be removed, making a lot of intra-doc links to not be inferred.
+    ///
+    /// So instead, we store the ambiguous links and we wait for cache paths to be filled before
+    /// inferring them (if possible).
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:296]: Can be done as a follow-up but if you have ideas for improvements, it's very welcome!

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 3]
```diff
@@ -2061,6 +2251,8 @@ fn ambiguity_error(
         // There is no way for users to disambiguate at this point, so better return the first
         // candidate and not show a warning.
         return false;
+    } else if !emit_error {
+        return true;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:2255]: Good idea!

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:2255]: Ah actually not really: we need the `kinds` variable that is computed at the beginning...

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:305]: So with:

```rust
//! [struct@X]

mod a {
    pub struct X;
}

mod b {
    pub struct X;
}

pub use a::*;
pub use b::*;
```

We get:

```
warning: public documentation for `bar` links to private item `X`
 --> bar.rs:1:6
  |
1 | //! [struct@X]
  |      ^^^^^^^^ this item is private
  |
  = note: this link will resolve properly if you pass `--document-private-items`
  = note: `#[warn(rustdoc::private_intra_doc_links)]` on by default
```

I think it's a bug but for now, let's simply remove this `disambiguator` as it doesn't seem to work.

**Comment by GuillaumeGomez** [comment]
Updated!

**Pull request merged**